### ISSUE TITLE
Ctrl-C not parsed correctly

### ISSUE THREAD

**Issue by battaglia01** [xonsh/xonsh#2599]
For xonsh 0.6.0, running on Python 3.4.3 on Windows within the Ubuntu bash shell, running a command such as "yes" causes xonsh to hang. The command "yes" simply outputs continuously its argument, so that you have

$ yes test
test
test
test
...

This repeats until you hit Ctrl-C.

In Xonsh, Ctrl-C has no effect, and it hangs.

**Comment by melund** [comment]
Can you post the output of running `xonfig`?

**Comment by melund** [comment]
So, I hear you say, the problem is that Ctrl-C is not intercepted when running in Linux version of Xonsh in WSL on windows?

**Comment by scopatz** [comment]
Thanks for reporting @battaglia01.  Can you report the message it prints when it crashes?

**Comment by gforsyth** [comment]
I ran into this a few days ago and meant to put in a fix, although it may only be a workaround.  Adding `yes` to the `threadable_predictors` will allow Ctrl-C to intercept (at least on Arch)

`__xonsh_commands_cache__.threadable_predictors['yes'] = lambda args: False`

**Comment by scopatz** [comment]
Hmm weird.  That isn't right.  I am also seeing this issue.  This really isn't the use case for threadable predictors, which is meant for shells in interactive mode, text editors, etc.  yes should totally be cancelable. This indicates that something (like a bare except) is eating keyboard interrupts.

**Comment by melund** [comment]
I can't reproduce the error directly on Windows. Ctrl-C works fine for me. I will try with within WSL when I get some time.

**Comment by battaglia01** [comment]
Also, I had a typo in my first post - xonsh doesn't "crash," it just hangs. Actually, to be more specific, it doesn't do anything. If you type "yes test", it keeps outputting "test" and Ctrl-C has no effect, so you have to close the terminal and restart.

Is threadable predictors supposed to be a whitelist of programs for which this works?

**Comment by battaglia01** [comment]
Quick note - I also have the problem on KDE Neon 5.12. Typing "yes" causes the same behavior. Here is my xonfig:

```
+------------------+----------------------+
| xonsh            | 0.6.0                |
| Git SHA          | c29819c5             |
| Commit Date      | Nov 22 17:51:52 2017 |
| Python           | 3.5.2                |
| PLY              | 3.9                  |
| have readline    | True                 |
| prompt toolkit   | None                 |
| shell type       | readline             |
| pygments         | None                 |
| on posix         | True                 |
| on linux         | True                 |
| distro           | neon                 |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```

**Comment by dosimple** [comment]
Actually, `yes` seems to be stopped just fine. But it takes a while show all the output from `yes`, even after is has been killed. So it looks like `yes` has hung. (On Linux)

**Comment by battaglia01** [comment]
I just checked on KDE Neon and have the same result as @dosimple -- Ctrl-C does (eventually) register, but it takes a lot of time for the output to catch up.

**Comment by battaglia01** [comment]
OK, some more info -- my problem description was wrong. It seems to be a concurrency issue.

Ctrl-C *DOES* register, but the output display blocks, so if there's a huge backlog built up then it takes forever for the Ctrl-C to register.

This is definitely not the right behavior -- we want to process Ctrl-C asynchronously and dump whatever output is remaining. Likewise for Ctrl-Z and other interrupts.

So perhaps this is related to the `threadable_predictors` thing after all. Where can I read more about the architecture there?

But, the solution can't be just to add more things to a whitelist, since we really want Ctrl-C to work for any command, including scripts people might create themselves, obscure shell commands downloaded from apt, whatever.

Here is a list of commands that exhibit the problem to varying degrees. In all cases, if you wait long enough, Ctrl-C will register:

- yes
- find /
- xxd [any enormous file]
- cat [any enormous file]
- tree / (a bit less so)

In contrast, bash on Windows does all of these correctly except for cat, which actually crashes bash here as well.

**Comment by battaglia01** [comment]
OK, some more results supporting the notion that it's a concurrency issue.

First, the problem also happens in pure python mode. For instance, you can type this and get the behavior:
```
$ print('hi\n'*100000)
hi
hi
...
[Ctrl-C takes forever to process]
```

But, interestingly, this is fine:
```
  $ for n in range(1,100000):
...     print('hi')
hi
hi
...
[Ctrl-C processed with no delay]
```

I think the problem is that xonsh interleaves Ctrl-C parsing synchronously in between calls to write to the output. So if you have a billion quick calls to output, Ctrl-C is parsed just fine. If you have one huge, giant call to output, then it takes forever for it to parse Ctrl-C, because it blocks, or something like that.

**Comment by gforsyth** [comment]
Hey @battaglia01 -- nice job tracking this down!
I think your diagnosis about the timing for when Python can interpret the `KeyboardInterrupt` is correct.  If it also occurs in pure Python, I'm not sure if we can get around that issue, but we can try to find something.

**Comment by scopatz** [comment]
I think the issue may be that the custom SIGINT signal handler we have is not actually sending the signal to the underlying process, but rather just closing the file handles....

**Comment by scopatz** [comment]
Great work on tracking this down @battaglia01!

**Comment by scopatz** [comment]
Actually it is sending the signal, but not closing the handles! Sorry I was looking at the proxy class, which runs functions as subprocs.

**Comment by scopatz** [comment]
Weird, it is like the signal handler isn't being called at all, I wonder if something is replacing it out from under us.

**Comment by battaglia01** [comment]
I note the Python shell also exhibits the same problem, so at least this issue isn't entirely xonsh-specific. My instinct would to set up a buffer to break up large writes, which would probably make message passing easier to handle in general, but I think I'd need to understand better the way xonsh does things to really suggest anything concrete...

Also, adding to threadable_predictors effectively fixes this?

**Comment by scopatz** [comment]
Setting threadable_predictors just avoids the issue, and doesn't really fix it.  There seems to be some weird interplay with signals/threading/forking/bound methods.  I have been able to verify that the correct sigint for the correct thread is registered as the signal handler.  However, when I check the `id()` bound method of the registered signal handler vs the actual bound method, they are different!  This is truly weird.  Maybe signal.signal() is copying bytecode or something weird?

**Comment by battaglia01** [comment]
Hm, alright - I was going to read up on the threadable_predictors stuff, but I guess that doesn't really apply here, so no point. I'll do some digging...

**Comment by scopatz** [comment]
oof - even this Python script reproduces the problem:

```python
import subprocess
p = subprocess.Popen(['yes'])
```

**Comment by scopatz** [comment]
Ok, it seems the xonsh is actually killing the process, but then it is spending eternity drawing down the buffered output.

**Comment by scopatz** [comment]
On my machine it is taking a little more than a minute to clear the buffer.

**Comment by scopatz** [comment]
This seems to be related to `$XONSH_PROC_FREQUENCY`

**Comment by scopatz** [comment]
Potential fix in #2616 - this one took a lot of debugging :)

**Comment by anki-code** [comment]
Can't reproduce this on xonsh 0.9.26:
```python
yes test
test
test
test
test
test
^C
```
